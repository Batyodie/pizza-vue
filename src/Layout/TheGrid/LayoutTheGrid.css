@import "../../style/Base/Extend/BaseExtend.css";

.TheGrid {
  @extend .ContentWrapper;
  flex-direction: column;
  justify-content: flex-start;
  align-items: flex-start;
}
.Title {
  @extend .BaseTitle;
}
.Layout {
  display: grid;
  grid-template-columns: 1fr 1fr 1fr 1fr;
  grid-template-rows: 1fr 1fr;
  grid-gap: 35px;
  grid-template-areas:
    ". . . ."
    ". . . .";
  align-items: self-end;
}
@media all and (-ms-high-contrast: none) {
  .Layout {
    display: -ms-grid;
    -ms-grid-columns: 1fr 1fr 1fr 1fr;
    -ms-grid-rows: 1fr 1fr;
    grid-gap: 35px;
  }
}

@media (max-width: 1400px) {
  .Layout {
    align-items: center;
    align-self: center;
    grid-template-columns: 1fr 1fr 1fr;
    grid-template-rows: auto;
    grid-template-areas:
      ". . . "
      ". . . "
      ". . .";
  }
}

@media (max-width: 1100px) {
  .Layout {
    grid-template-columns: 1fr 1fr;
    grid-template-rows: auto;
    grid-template-areas:
      ". . "
      ". . "
      ". . ";
  }
}

@media (max-width: 810px) {
  .Layout {
    grid-template-columns: 1fr;
    grid-template-rows: auto;
    grid-template-areas:
      ". "
      ". "
      ". ";
  }
}

@media (max-width: 810px) {
  .TheGrid {
    padding: 20px 10px;
  }
}

@media (max-width: 425px) {
  .TheGrid {
    padding: 10px 10px;
  }
}
