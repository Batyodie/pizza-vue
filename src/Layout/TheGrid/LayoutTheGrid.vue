<script>
import { Card } from "@/components";
import { ContentLoader } from "vue-content-loader";
import { mapGetters, mapActions } from "vuex";

import { TheGridStyle } from "@/Layout/style";
export default {
  name: "LayoutTheGrid",
  components: { Card, ContentLoader },
  data() {
    return {};
  },

  computed: {
    TheGridStyle() {
      return TheGridStyle;
    },

    ...mapGetters({
      cards: "getCards",
      cart: "getCart",
      getCartItem: "getCartItem",
      isLoaded: "getPizzasLoadedFlag"
    }),
    getLastCardTags() {
      return id => (this.getCartItem(id) ? this.getCartItem(id).tags : null);
    }
  },

  methods: {
    ...mapActions({
      addPizza: "addPizzaToCart"
    }),

    onClickAddPizza(obj) {
      this.addPizza(obj);
    }
  }
};
</script>

<template>
  <main :class="TheGridStyle.TheGrid">
    <div style="margin-bottom: 35px;">
      <h1 :class="TheGridStyle.Title">Все пиццы</h1>
    </div>
    <div :class="TheGridStyle.Layout">
      <template v-if="!isLoaded">
        <content-loader
          style="min-width: 280px"
          :width="284"
          :height="523"
          :speed="2"
          v-for="(i, _i) in cards.length || 12"
          :key="_i"
          primaryColor="#f3f3f3"
          secondaryColor="#ecebeb"
        >
          <circle cx="140" cy="140" r="140" />
          <rect x="6" y="299" rx="10" ry="10" width="270" height="40" />
          <rect x="6" y="354" rx="10" ry="10" width="270" height="104" />
          <rect x="6" y="491" rx="10" ry="10" width="96" height="27" />
          <rect x="126" y="476" rx="30" ry="30" width="153" height="44" />
        </content-loader>
      </template>
      <template v-else>
        <Card
          v-for="card in cards"
          :key="card.id"
          :onClickAddPizza="onClickAddPizza"
          :card="card"
          :cardIndex="card.id"
          :tags="getLastCardTags(card.id)"
        >
        </Card>
      </template>
    </div>
  </main>
</template>
