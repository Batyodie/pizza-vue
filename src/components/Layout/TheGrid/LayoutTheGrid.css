@import "../../Base/Styles/BaseExtend.css";

.TheGrid {
  @extend .ContentWrapper;
  flex-direction: column;
  justify-content: flex-start;
  align-items: flex-start;
}
.TheGridTitle {
  @extend .BaseTitle;
}
.TheGridLayout {
  display: grid;
  grid-template-columns: 1fr 1fr 1fr 1fr;
  grid-template-rows: 1fr 1fr;
  grid-gap: 35px;
  grid-template-areas:
    ". . . ."
    ". . . .";
}
@media all and (-ms-high-contrast: none) {
  .TheGridLayout {
    display: -ms-grid;
    -ms-grid-columns: 1fr 1fr 1fr 1fr;
    -ms-grid-rows: 1fr 1fr;
    grid-gap: 35px;
  }
}

@media (max-width: 1400px) {
  .TheGridLayout {
    align-items: center;
    align-self: center;
    grid-template-columns: 1fr 1fr 1fr;
    grid-template-rows: auto;
    grid-template-areas:
      ". . . "
      ". . . "
      ". . .";
  }
}

@media (max-width: 1100px) {
  .TheGridLayout {
    grid-template-columns: 1fr 1fr;
    grid-template-rows: auto;
    grid-template-areas:
      ". . "
      ". . "
      ". . ";
  }
}

@media (max-width: 810px) {
  .TheGridLayout {
    grid-template-columns: 1fr;
    grid-template-rows: auto;
    grid-template-areas:
      ". "
      ". "
      ". ";
  }
}

@media (max-width: 810px) {
  .TheGrid {
    padding: 20px 30px;
  }
}

@media (max-width: 425px) {
  .TheGrid {
    padding: 10px 10px;
  }
}
