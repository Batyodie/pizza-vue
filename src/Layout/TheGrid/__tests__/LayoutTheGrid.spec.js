import { createLocalVue, mount, shallowMount } from "@vue/test-utils";
import Vuex from "vuex";

import LayoutTheGrid from "@/Layout/TheGrid/LayoutTheGrid";

const localVue = createLocalVue();
localVue.use(Vuex);

describe("Unit test from LayoutTheGrid components", () => {
  // fake data
  const products = [
    {
      id: 7,
      imageUrl:
        "https://dodopizza.azureedge.net/static/Img/Products/Pizza/ru-RU/d48003cd-902c-420d-9f28-92d9dc5f73b4.jpg",
      name: "Маргарита",
      types: [0, 1],
      sizes: [26, 30],
      price: 450,
      category: 4,
      rating: 10,
    },
    {
      id: 8,
      imageUrl:
        "https://dodopizza.azureedge.net/static/Img/Products/Pizza/ru-RU/ec29465e-606b-4a04-a03e-da3940d37e0e.jpg",
      name: "Четыре сезона",
      types: [0],
      sizes: [26, 40],
      price: 395,
      category: 5,
      rating: 10,
    },
    {
      id: 6,
      imageUrl:
        "https://dodopizza.azureedge.net/static/Img/Products/Pizza/ru-RU/d2e337e9-e07a-4199-9cc1-501cc44cb8f8.jpg",
      name: "Пепперони",
      types: [1],
      sizes: [30, 40],
      price: 675,
      category: 1,
      rating: 9,
    },
  ];
  const loadedProducts = true;

  let store;
  let actions;
  let getters;
  beforeEach(() => {
    actions = {
      addPizzaToCart: jest.fn(),
    };
    getters = {
      getCards: () => products,
      getPizzasLoadedFlag: () => loadedProducts,
      getCartItem: () => () => null,
      getCartItemType: () => () => null,
    };

    store = new Vuex.Store({
      actions,
      getters,
    });
  });
  // create wrapper factory
  const wrapperFactory = (mouning = shallowMount, { propsData } = {}) => {
    return mouning(LayoutTheGrid, {
      localVue,
      store,
      propsData: {
        ...propsData,
      },
    });
  };
  it("should initialize correctly layoutTheGrid component", () => {
    const wrapper = wrapperFactory(mount);
    const cards = wrapper.findAll(".Card");

    // correctly Vue instance
    expect(wrapper.exists()).toBeTruthy();
    // We expect the wrapper to be layout the grid component
    expect(wrapper.is(LayoutTheGrid)).toBeTruthy();
    expect(cards.length).toBe(3);
  });
  it("should correctly render cards in  layoutTheGrid component", () => {
    const wrapper = wrapperFactory(mount);
    let index = 0;
    const cards = wrapper.findAll(".Card");

    cards.wrappers.forEach((e, i) => {
      const renderCardTitle = `${products[index].name}`;
      const renderCardImg = products[index].imageUrl;
      const renderCardPrice = `от ${products[index].price} ₽`;

      if (index === i) {
        const element = cards.at(index);
        expect(element.find(".Title").text()).toEqual(renderCardTitle);
        expect(element.find(".Image > img").attributes("src")).toEqual(
          renderCardImg
        );
        expect(element.find(".PizzaPrice").text()).toEqual(renderCardPrice);
        expect(
          element
            .findAll(".TagsItem")
            .at(products[index].types[0])
            .classes("TagsItemActive")
        ).toEqual(true);
        index++;
      }
    });
    const cardSize = (cardIndex, tagIndex) =>
      cards
        .at(cardIndex)
        .findAll(".TagsItem")
        .at(tagIndex);

    expect(cardSize(0, 2).classes("TagsItemActive")).toBe(true);
    expect(cardSize(1, 2).classes("TagsItemActive")).toBe(true);
    expect(cardSize(2, 2).classes("TagsItemActive")).toBe(false);
  });

  it("should match the snapshot layoutTheGrid component", () => {
    const wrapper = wrapperFactory(mount);
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("should correctly add pizza item to card", async () => {
    const wrapper = wrapperFactory(mount);
    const card = wrapper.findAll(".Card");
    const addBtn = card.at(0).find(".AddBtn");

    await addBtn.trigger("click");
    expect(actions.addPizzaToCart).toHaveBeenCalled();
  });
});
