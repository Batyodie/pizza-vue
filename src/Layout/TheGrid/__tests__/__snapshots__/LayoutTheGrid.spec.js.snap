// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unit test from LayoutTheGrid components should match the snapshot layoutTheGrid component 1`] = `
<main class="TheGrid">
  <div style="margin-bottom: 35px;">
    <h1 class="Title">Все пиццы</h1>
  </div>
  <div class="Layout">
    <div class="Card">
      <picture class="Image">
        <source> <img src="https://dodopizza.azureedge.net/static/Img/Products/Pizza/ru-RU/d48003cd-902c-420d-9f28-92d9dc5f73b4.jpg" alt="Pizza">
      </picture>
      <h3 class="Title">
        Маргарита
      </h3>
      <div class="Tags"><button type="button" class="Btn TagsItemActive TagsItem Tag"><span class="Body"> тонкое </span> </button><button type="button" class="Btn TagsItem Tag"><span class="Body"> традиционное </span> </button> <button type="button" class="Btn TagsItemActive TagsItem Tag"><span class="Body"> 26 см. </span> </button><button type="button" class="Btn TagsItem Tag"><span class="Body"> 30 см. </span> </button><button type="button" class="Btn TagsDisable TagsItem Tag"><span class="Body"> 40 см. </span> </button></div>
      <div class="PizzaAdd">
        <h3 class="PizzaPrice">от 450 ₽</h3> <button type="button" class="Btn AddBtn"><span class="Body">
          Добавить
        <!----></span> <svg width="12" height="12" viewBox="0 0 12 12" fill="none" xmlns="http://www.w3.org/2000/svg" class="AddBtnIcon">
            <path d="M10.8 4.8H7.2V1.2C7.2 0.5373 6.6627 0 6 0C5.3373 0 4.8 0.5373 4.8 1.2V4.8H1.2C0.5373 4.8 0 5.3373 0 6C0 6.6627 0.5373 7.2 1.2 7.2H4.8V10.8C4.8 11.4627 5.3373 12 6 12C6.6627 12 7.2 11.4627 7.2 10.8V7.2H10.8C11.4627 7.2 12 6.6627 12 6C12 5.3373 11.4627 4.8 10.8 4.8Z" fill="white"></path>
          </svg></button>
      </div>
    </div>
    <div class="Card">
      <picture class="Image">
        <source> <img src="https://dodopizza.azureedge.net/static/Img/Products/Pizza/ru-RU/ec29465e-606b-4a04-a03e-da3940d37e0e.jpg" alt="Pizza">
      </picture>
      <h3 class="Title">
        Четыре сезона
      </h3>
      <div class="Tags"><button type="button" class="Btn TagsItemActive TagsItem Tag"><span class="Body"> тонкое </span> </button><button type="button" class="Btn TagsDisable TagsItem Tag"><span class="Body"> традиционное </span> </button> <button type="button" class="Btn TagsItemActive TagsItem Tag"><span class="Body"> 26 см. </span> </button><button type="button" class="Btn TagsDisable TagsItem Tag"><span class="Body"> 30 см. </span> </button><button type="button" class="Btn TagsItem Tag"><span class="Body"> 40 см. </span> </button></div>
      <div class="PizzaAdd">
        <h3 class="PizzaPrice">от 395 ₽</h3> <button type="button" class="Btn AddBtn"><span class="Body">
          Добавить
        <!----></span> <svg width="12" height="12" viewBox="0 0 12 12" fill="none" xmlns="http://www.w3.org/2000/svg" class="AddBtnIcon">
            <path d="M10.8 4.8H7.2V1.2C7.2 0.5373 6.6627 0 6 0C5.3373 0 4.8 0.5373 4.8 1.2V4.8H1.2C0.5373 4.8 0 5.3373 0 6C0 6.6627 0.5373 7.2 1.2 7.2H4.8V10.8C4.8 11.4627 5.3373 12 6 12C6.6627 12 7.2 11.4627 7.2 10.8V7.2H10.8C11.4627 7.2 12 6.6627 12 6C12 5.3373 11.4627 4.8 10.8 4.8Z" fill="white"></path>
          </svg></button>
      </div>
    </div>
    <div class="Card">
      <picture class="Image">
        <source> <img src="https://dodopizza.azureedge.net/static/Img/Products/Pizza/ru-RU/d2e337e9-e07a-4199-9cc1-501cc44cb8f8.jpg" alt="Pizza">
      </picture>
      <h3 class="Title">
        Пепперони
      </h3>
      <div class="Tags"><button type="button" class="Btn TagsDisable TagsItem Tag"><span class="Body"> тонкое </span> </button><button type="button" class="Btn TagsItemActive TagsItem Tag"><span class="Body"> традиционное </span> </button> <button type="button" class="Btn TagsDisable TagsItem Tag"><span class="Body"> 26 см. </span> </button><button type="button" class="Btn TagsItemActive TagsItem Tag"><span class="Body"> 30 см. </span> </button><button type="button" class="Btn TagsItem Tag"><span class="Body"> 40 см. </span> </button></div>
      <div class="PizzaAdd">
        <h3 class="PizzaPrice">от 675 ₽</h3> <button type="button" class="Btn AddBtn"><span class="Body">
          Добавить
        <!----></span> <svg width="12" height="12" viewBox="0 0 12 12" fill="none" xmlns="http://www.w3.org/2000/svg" class="AddBtnIcon">
            <path d="M10.8 4.8H7.2V1.2C7.2 0.5373 6.6627 0 6 0C5.3373 0 4.8 0.5373 4.8 1.2V4.8H1.2C0.5373 4.8 0 5.3373 0 6C0 6.6627 0.5373 7.2 1.2 7.2H4.8V10.8C4.8 11.4627 5.3373 12 6 12C6.6627 12 7.2 11.4627 7.2 10.8V7.2H10.8C11.4627 7.2 12 6.6627 12 6C12 5.3373 11.4627 4.8 10.8 4.8Z" fill="white"></path>
          </svg></button>
      </div>
    </div>
  </div>
</main>
`;
